[Pytorch][quant] Move xnnpack quantizer to use aten.linear

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109254

Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

kimishpatel committed:
[Pytorch][quant] Move xnnpack quantizer to use aten.linear

Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -168,83 +166,45 @@ def _annotate_linear(
+            continue
+        if filter_fn and not filter_fn(node):
+            continue
+        annotated_partitions.append([node])
```
since you are modifying this, this should happen after the is_annotated check in L191 I think

kimishpatel replied:
Oh thats right. Let me move.

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -168,83 +166,45 @@ def _annotate_linear(
-            _mark_nodes_as_annotated(nodes_to_mark_annotated)
+            _annotate_output_qspec(node, output_act_qspec)
+
+        if bias_node and _is_annotated([bias_node]) is False:
```
why are we annotating output of bias? shouldn't it be input to aten.linear?

kimishpatel replied:
yeah i just forgot to update this. It can also be annotated as input.

jerryzh168 reviewed (changes requested):
thanks, this is much better. had some comments inline

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -168,83 +166,45 @@ def _annotate_linear(
+        act_node = node.args[0]
+        weight_node = node.args[1]
+        bias_node = None
+        if len(node.args) < 2:
```
is this enforced by the schema for aten.linear op itself?

kimishpatel replied:
you cant have aten.linear without input + weight. https://fburl.com/code/knz9927g

jerryzh168 replied:
right, I mean we don't need to check for this since it's enforced by the schema?

kimishpatel replied:
oh i see what you mean. yes dont need to.

jerryzh168 replied:
are you planning to remove this?

kimishpatel replied:
oh sorry forgot

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -168,83 +166,45 @@ def _annotate_linear(
+            )
+        if len(node.args) > 2:
+            bias_node = node.args[2]
+        # find use of act node within the matched pattern
```
nit: remove

kimishpatel committed:
Update on "[Pytorch][quant] Move xnnpack quantizer to use aten.linear"


Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -168,83 +166,53 @@ def _annotate_linear(
+                weight_qspec,
+            )
+            nodes_to_mark_annotated = [node, weight_node]
+            if bias_node and _is_annotated([bias_node]):
```
I feel we don't need this check?

kimishpatel replied:
why not?

kimishpatel replied:
In fact I should check the same for weight node as well

jerryzh168 replied:
I think annotating bias is still valid?

kimishpatel replied:
are you saying that as an edge annotation it is still valid to annotate? Yeah that makes sense, although as a parameter on linear, do we want that? this allows us associating one linear with one weight and one bias. 

What you are saying is also valid, but I am erring on the side of caution for now to not allow this behavior and introduce inadvertent behaviort

jerryzh168 replied:
quantizing output for bias_node won't interfere with the annotation for bias input for linear node I think since if it does not match the linear config, we'll insert a new observer.

quantizing input for bias_node is not related to the linear annotation, so I don't think it should be banned here.

I guess erroring out is fine as well since most of the time these are get_attr nodes and not shared with other ops, just feels like it's not needed. it might break in some cases but it's fine to adjust later as well.

kimishpatel replied:
I think you are right. Lets just allow the use annotation. It does have side effect when there are multiple uses of weight and bias, although very unlikely, but it is ok for quantizer to not be opinionated about it.

kimishpatel committed:
Update on "[Pytorch][quant] Move xnnpack quantizer to use aten.linear"


Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).

kimishpatel committed:
Update on "[Pytorch][quant] Move xnnpack quantizer to use aten.linear"


Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).

kimishpatel committed:
Update on "[Pytorch][quant] Move xnnpack quantizer to use aten.linear"


Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).

jerryzh168 reviewed (approved):
thanks for addressing all the comments! this looks pretty clean

kimishpatel committed:
Update on "[Pytorch][quant] Move xnnpack quantizer to use aten.linear"


Summary:
Now that quantization works on pre-dispatch aten IR, moving to full set
of aten ops is ok. Plus when tracing models like ViT, the linear
projections of of k, q, v uses functional.linear and not nn.Linear,
which results not being able to extract nodes corresponding to linear.

Test Plan:
quant tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D49252194](https://our.internmc.facebook.com/intern/diff/D49252194)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49252194).
